### DISCUSSION TITLE
Strugling with Layouts

### DISCUSSION THREAD

**Discussion by kirildi** [DioxusLabs/dioxus#1528] [Q&A] [answered]
Hello, I'm recently started a new project with Dioxus 0.4.1 FullStack, and found that it have the feature "Layouts". Which allows 
some components to persist while changing the page via link (if I understand it correctly)
My problem is that I'm trying to use `Outlet::<Route>` inside html 'main' component, and it builds and runs the server successfully
but when I try to access the address it prints in the console 
``` 
thread 'tokio-runtime-worker' panicked at 'Outlet must be inside of a router' 
thread 'tokio-runtime-worker' panicked at 'not yet implemented: generally, scopes should be sync, only if being traversed'
thread 'tokio-runtime-worker' panicked at 'called `Result::unwrap()` on an `Err` value: RecvError(())'
```
 my app.rs is
``` 
pub fn App(cx: Scope) -> Element {    
    cx.render(rsx! {
        div{
            Header {},
            Main {}
        }
    })
}
```

and my Main.rs is:
```
#[derive(Routable, Clone)]
#[rustfmt::skip]
pub enum Route {
    #[layout(Main)]
        #[route("/")]
        Main {},
}

#[inline_props]
pub fn Main(cx: Scope) -> Element {
    render(rsx! {
        main {
            Outlet::<Route> {}
        }
    })
}
```

Thanks in advance for the help,

**Comment by ealmloff** [comment, marked as answer]
Outlets tell Layouts where to renderer child routes (any routes under the layout). Only layout components should have outlets in them and layout components need to be rendered by the router.

In your app, the Header component should be a layout component with an Outlet component placed inside of that component where you want to render the main part of your page.

To render the Router, you can use the router component, not a individual page from the router.


app.rs:
```rust
pub fn App(cx: Scope) -> Element {    
    cx.render(rsx! {
        div{
            Router::<Route> {}
        }
    })
}
```
and my Main.rs is:

```rust
#[derive(Routable, Clone)]
#[rustfmt::skip]
pub enum Route {
    #[layout(Header)]
        #[route("/")]
        Main {},
}

#[inline_props]
pub fn Main(cx: Scope) -> Element {
    render(rsx! {
        main {
            "This is the home page"
        }
    })
}

#[inline_props]
pub fn Header(cx: Scope) -> Element {
    render(rsx! {
         // Header elements here
         div {}
         // This renderers the Home component (or another active component inside of the layout)
         Outlet::<Route> {}
    })
}
```

**Reply by kirildi** [reply]
Thank you. I definitely missed that I need Router::Route first  👍🏽
